S matrix tool incorporates element mapping and row-wise computation

tylerflex opened the pull request:
incorporates the following additions and changes to `ComponentModeler`:

```python
MatrixIndex = Tuple[Port, pd.NonNegativeInt]
Element = Tuple[MatrixIndex, MatrixIndex]
SMatrixType = Dict[Element[Dict[Element, complex]]

class ComponentModeler(Tidy3dBaseModel):

    ...

    element_mappings: Dict[Element, Tuple[Element, bool]] = pd.Field(
        {},
        title="Element Mappings",
        description="Mapping between matrix indices of the scattering matrix, "
        "specified by (:class:`.Port`, `int`). "
        "``element_mappings(port, mode_index)`` returns ``(port, mode_index), pos_sign``, "
        "where the ``(port, mode_index)`` refers to the output matrix index "
        "and ``pos_sign`` determines whether the element should be copied with a positive sign.",
    )
    run_only: Optional[Tuple[MatrixIndex, ...]] = pd.Field(
        None,
        title="Run Only",
        description="If specified, a tuple of matrix indices, specified by (:class:`.Port`, `int`),"
        " to run only, excluding the other rows from the scattering matrix. "
        "If this option is used, the resulting scattering matrix will not be square or complete.",
    )
```

* matrix indices are specified by `(port, mode_index)` pairs
* scattering matrix `S` is indexed by `S[(port_in, mode_index_in)][(port_out, mode_index_out)] -> complex`.
* user can specify tuple of `(port, mode_index)` to "run only". These rows omitted from the scattering matrix.
* user can specify dict of `((port, mode_index), (port, mode_index)) -> ((port, mode_index), (port, mode_index)), bool` mapping s_matrix elements to another. If the boolean value is `False`, it will apply negative sign in the mapping.
* for a given `(port, mode_index)` specifying a source, if all `(port, mode_index)` outputs are "covered" by such mappings from other sources, it is not solved for with a simulation (what about self-self interactions? do we need to map these)?

Note, have not run because develop needs backend working first.

momchil-flex commented on tidy3d/plugins/smatrix/smatrix.py:
```diff
@@ -92,31 +95,72 @@ class ComponentModeler(Tidy3dBaseModel):
     )
-
-    batch: Tidy3dBaseModel = pd.Field(
+    element_mappings: Dict[Element, Tuple[Element, bool]] = pd.Field(
```
Yeah this is general enough alright but certainly not super convenient. This mapping can also be constructed by an isomorphism relationship like (p0, p1, p2) (p0, p2, p1) and an `invert_sign` list of `(Port, mode_index)` corresponding to that particular isomorphism. E.g. in the splitter case, the list will contain all the modes of port 0 that are antisymmetric. Maybe we can have a function that takes input like that and returns an `element_mappings` dict to be fed in here?

tylerflex replied:
I'm not sure exactly how the `invert_sign` is handled. So if `invert_sign = [p1, 0]` does that mean that any element mappings from `p0, 0` to *any* other port should have an inverted sign? and everything else should not have an inverted sign? otherwise, I agree that an alternative constructor like this would be useful.

momchil-flex replied:
> I'm not sure exactly how the `invert_sign` is handled. So if `invert_sign = [(p0, 0)]` does that mean that any element mappings from `(p0, 0)` to _any_ other port should have an inverted sign? and everything else should not have an inverted sign?

I think it needs to be applied whenever the isomorphism enforces something like `S_ij = S_iq` or `S_ji = S_qi`, where `i` is the mode that has "invert_sign" under that isomorphism.

I'm fine with putting this off for now though.

tylerflex commented on tidy3d/plugins/smatrix/smatrix.py:
```diff
@@ -63,7 +64,9 @@ def evaluate_mode_indices(cls, val, values):
 relating the coupling amplitudes between the m and n mode orders in the two ports, respectively.
 """
-SMatrixType = Dict[str, np.ndarray]
+MatrixIndex = Tuple[Port, pd.NonNegativeInt]
```
we can do `str` for name, makes sense.

tylerflex commented:
ok, @momchil-flex I think this is just about ready to merge now.

I added a lot more testing, including of new features. Verified that the s matrices returned by various setups have the expected elements (and the ones that are not expected are not present).  Also verified that simulations will not be run if they are covered by mappings or present in `run_only`.  I moved all of these tests to their own file `test_component_modeler.py` because the plugins test was getting too long.

Also updated the notebook a bit and verified that it runs and edited the text a bit. I have not put any examples of the mappings but maybe we can put them in if there's a good example.

in favor of putting off the permutation business for now, we'd just need to write a `class method` constructor that generates the `element_mapping` dict from a given set of permutations and sign inversions. Everything else should work as expected once that is generated. I'm a bit fuzzy on the details but can help with the implementation when we get to it.

There is one remaining issue, saving and loading the `ComponentModeler` to and from file seems to erase the data. I wasn't able to find a great workaround but here's the issue in its essence:

the `ComponentModeler.batch` is a `@cached_property` that initializes a `Batch()` object using the simulations needed to create the S matrix.

the `ComponentModeler.run(path_dir)` runs the batch, and constructs the S matrix from the `batch_data`.

If we save the `ComponentModeler` to file, the `batch` does not get stored because it's a `@cached_property`.

When we load the `ComponentModeler` from file and try to load the batch data from its batch, 
```py
c = ComponentModeler.from_file(path)
batch_data = c.batch.load()
```
the call to `c.batch` initializes a *new* `Batch`, which will generate a new set of `jobs` with different task_ids and upload them as drafts.

I looked into a few solutions for this but couldn't come up with anything great:
1. manually save the batch to file by overloading `ComponentModeler.to_file()` and try to load it in `self.batch` property. Got really messy.
2. make `batch` a field, but then it needs the simulations so it gets really messy again.
3. maybe we should try to save `cached_properties` to file? although this makes a bit of a mess because then the files will contain a ton of extraneous information.

Any thoughts?

momchil-flex commented:
> To summarize the problem a bit more generally: If the batch is a property, it gets regenerated when we load from file. if the batch is a field, we need to do most of the s matrix logic (setting up simulations) in a validators, which goes against a lot of the design principles and is pretty ugly.
> 
> I looked into a few solutions for this but couldn't come up with anything great:
> 
> 1. manually save the batch to file by overloading `ComponentModeler.to_file()` and try to load it in `self.batch` property. Got really messy.
> 2. make `batch` a field, but then it needs the simulations so it gets really messy again.
> 3. maybe we should try to save `cached_properties` to file? although this makes a bit of a mess because then the files will contain a ton of extraneous information.
> 
> Any thoughts?

So another potential approach around this is to *use* the fact that simulations on the server can be updated before they are run (this is related to our discussion of whether they should be stored in Job/Batch at all in that case). Consider this:

- Make batch a Field again
- Upon initialization, in the `self.batch` validator use `sim_dict = {task_name: empty_sim}` where the `task_name`-s are correct for the current S-matrix, but `empty_sim` can be.. anything really.
- Upon `run`, you can first iterate over the actual `self.sim_dict` and update the remote `simulation.json` for each task. This is just a call to the `upload_file` function from the webapi, using the task_ids that have already been created and stored upon init.

Now, this is not perfect, but I don't know how it compares to the other possible solutions.

More generally, our `web.upload` function (and the corresponding wrappers in Job and Batch) really does two separate things: create a task with a given task_id (no simulation needed), and then upload a simulation to that task container. I think these could be split in two separate functions, at least internally, especially now that the part that purely does the simulation upload is a bit more complicated (potentially uploading extra data as hdf5). 

This is related to, but doesn't necessarily force, what we discussed earlier: does it make sense for `Job` and `Batch` to store the actual simulations, or should we make them just containers for task_ids? This certainly makes some sense for me for `Job`, but it is a bit more awkward for `Batch`, because on the surface it would look like you would have to pass e.g. a list of `num_tasks` task names upon init, and then a list of exactly the same length of simulations upon `batch.run`. But.. I don't know, it's still something to seriously consider I think.

tylerflex commented:
Interesting ideas. My feeling however is that this goes back a bit to our mutability vs immutability discussion, except now it's extended to the web UI. Allowing the simulation associated with a task ID to be modified is a bit like introducing mutability back into the entire workflow from `Simulation` -> `SimulationData`. I think it complicates things to allow this mutability and I don't see an obvious advantage to being able to do it. Although you propose some interesting ways that we could change our web API to support it, it still worries me that we will then need to keep track of the *state* of the simulation / task and it could complicate things or lead to annoying bugs.

The way I would imagine things working is that a "task" is something that is associated with a *single* Simulation and if you want to edit the simulation, you must `.copy()` and update the simulation and tag it with a new `task_id`. In this framework, perhaps even uploading the simulation as a draft should generate a task_id for it. We could generalize the GUI to say that if editing *any* simulation associated with a task, make a copy first.

For now, I will make the `batch` a `Field` of the `ComponentModeler` again and just try to revert to initializing it through the validator. For example, if the `batch` Field is supplied (ie loaded from file), use the supplied value, otherwise, generate all the necessary simulations, make a new batch. 

Problem with this approach is that if the user creates a `ComponentModeler` and wants to change some aspect of it, the newly created batch will have already uploaded a bunch of `tasks` to the server as draft. So each `.copy()` or re-running of the notebook with different parameters will upload a new set of tasks.. In the old approach, the tasks are only created once the `batch` is asked for (ie in `run()`)

tylerflex commented:
Ok, I made the change. Last thing I'm wondering is whether `element_mappings` should be a bit more general for example, instead of a `bool` to indicate sign flip, a function that can specify any mapping between the two elements
```py
element_mappings: Dict[Element, Tuple[Element, Callable[[complex], complex]]
```

Then one could supply some operations like
```py
lambda x: x
lambda x: -x
lambda x: np.conj(x)
lambda x: 2*x
```

do you think there's a use case for this? if so, might as well just put it in now, I suppose.

momchil-flex commented:
> Interesting ideas. My feeling however is that this goes back a bit to our mutability vs immutability discussion, except now it's extended to the web UI. Allowing the simulation associated with a task ID to be modified is a bit like introducing mutability back into the entire workflow from `Simulation` -> `SimulationData`. I think it complicates things to allow this mutability and I don't see an obvious advantage to being able to do it. Although you propose some interesting ways that we could change our web API to support it, it still worries me that we will then need to keep track of the _state_ of the simulation / task and it could complicate things or lead to annoying bugs.
> 
> The way I would imagine things working is that a "task" is something that is associated with a _single_ Simulation that has been uploaded to the server. If you want to edit a simulation associated with a task, you must `.copy()` and update the simulation and tag it with a new `task_id`. In this framework, uploading the simulation as a draft should generate a `task_id` for it. For the logic required by the GUI, we can just say that editing _any_ simulation associated with a task, make a copy first. To make this logic work out, we'd need to enforce that simulations being edited by the GUI are not saved as "draft" but some other status that does not have a task_id (just a reference to the json file storing the simulation state).

I think this breaks a lot of how web stuff currently works and may not be easy to convince MC to do it this way overall. In some sense, the whole idea of the "draft" status is exactly that the simulation can still be edited. And I think something like a task_id is required because the data associated to the draft (e.g. the simulation json) needs to already be stored somewhere on S3. So basically they'll need to introduce a non-task_id task_id, and a non-"draft" status which essentially also means "draft" (or rename the current "draft" status to something like "uploaded"?)

But anyway, once it's available, we could also just turn on the web "lock" that I requested for every simulation that's uploaded from python, while leaving the GUI workflow untouched?

momchil-flex commented:
> Ok, I made the change. Last thing I'm wondering is whether `element_mappings` should be a bit more general for example, instead of a `bool` to indicate sign flip, a function that can specify any mapping between the two elements
> 
> ```python
> element_mappings: Dict[Element, Tuple[Element, Callable[[complex], complex]]
> ```
> 
> Then one could supply some operations like
> 
> ```python
> lambda x: x
> lambda x: -x
> lambda x: np.conj(x)
> lambda x: 2*x
> ```
> 
> do you think there's a use case for this? if so, might as well just put it in now, I suppose.

I am not sure... I like the generality, but it's less convenient for simple things. However, overall `element_mappings` already looks not that convenient to use by hand. So maybe we introduce this more general approach there and in the future think of convenience methods?

tylerflex commented:
Ok pushed the more general mapping stuff.

I'm not really sure what to do about the task mutability issue. Seems we can't save and load a batch (which just contains jobs, which just contain task_ids and simulations) with 100% confidence that the simulations in the loaded batch match those on the server. I guess the only solution (besides making tasks "immutable" would be to use properties for the simulation that call the web, I think you proposed this before. For now, let's just assume that the locking mechanism will be in place in the future and not worry about it for now?

momchil-flex commented on tidy3d/plugins/smatrix/smatrix.py:
```diff
@@ -36,7 +36,7 @@ class Port(Box):
         description="Specifies how the mode solver will solve for the modes of the port.",
     )
-    mode_indices: Tuple[pd.NonNegativeInt, ...] = pd.Field(
+    mode_indices: Optional[Tuple[pd.NonNegativeInt, ...]] = pd.Field(
```
I wonder if we should remove this for simplicity? We just run all modes in the ModeSpec. If a user wants to only *source* a subset of the modes, we now have the `run_only` parameter. If they want to use modes that are not in order, e.g. 0 and 2, they can just add two ports.

One thing that is a bit confusing at first and made me think about this is that I was wondering if `MatrixIndex` indexes into the ModeSpec-defined array of indexes (`np.arange(mode_spec.num_modes)`) or the `Port.mode_indeces` subset of that array. I assume the latter.

tylerflex replied:
agreed, removed this.

momchil-flex commented on tests/utils.py:
```diff
@@ -235,6 +235,52 @@ def prepend_tmp(path):
+        coords = {"f": list(monitor.freqs)}
+        rmin, rmax = simulation.bounds
+        for dim, pos_min, pos_max in zip("xyz", rmin, rmax):
+            coords[dim] = np.linspace(pos_min, pos_max, 101)
```
For this to be consistent with what will actually be returned by the solver, the `coords` should be given by something like `simulation.discretize(monitor, extend=True).yee.E.x` for the Ex component, etc.

Well, there are some further considerations:
- if the monitor size is 0 along a certain dimension, the coord is just the center along that axis
- some monitors allow for spatial downsampling (record every `interval` spatial steps)

This is making me wonder if we shouldn't have something like a `Monitor.discretize(grid)` method instead of handling those things on the backend.

But yeah just generally it would be good to have this function return data with shapes matching what will actually be returned by the solver as closely as possible.

tylerflex replied:
yea, I omitted this out of laziness, but planned to add it in if this function sees more use.  I've added it in now:
```py
    def make_field_data(monitor: FieldMonitor) -> FieldData:
        """make a random FieldData from a FieldMonitor."""
        field_cmps = {}
        coords = {"f": list(monitor.freqs)}
        rmin, rmax = simulation.bounds
        for dim, pos_min, pos_max in zip("xyz", rmin, rmax):
            coords[dim] = np.linspace(pos_min, pos_max, 101)

        grid = simulation.discretize(monitor, extend=True)

        for field_name in monitor.fields:
            spatial_coords_dict = grid[field_name].dict()

            for axis, dim in enumerate('xyz'):
                if monitor.size[axis] == 0:
                    coords[dim] = [monitor.center[axis]]
                else:
                    coords[dim] = np.array(spatial_coords_dict[dim])

            field_cmps[field_name] = make_data(
                coords=coords, data_array_type=ScalarFieldDataArray, is_complex=True
            )

        return FieldData(monitor=monitor, **field_cmps)
```

momchil-flex commented on tidy3d/plugins/smatrix/smatrix.py:
```diff
@@ -92,32 +90,141 @@ class ComponentModeler(Tidy3dBaseModel):
+        {},
+        title="Element Mappings",
+        description="Mapping between elements of the scattering matrix, as specified by"
+        "specified by pairs of ``(port name, mode index)`` matrix indices. "
```
"specified by" repeated

momchil-flex commented on tidy3d/plugins/smatrix/smatrix.py:
```diff
@@ -92,32 +90,141 @@ class ComponentModeler(Tidy3dBaseModel):
+    run_only: Optional[Tuple[MatrixIndex, ...]] = pd.Field(
+        None,
+        title="Run Only",
+        description="If specified, a tuple of matrix indices, specified by (:class:`.Port`, `int`),"
```
``int``?

I guess we need to try to keep track of this / check in PRs: in docstrings, RST. In error messages - simple ' '

tylerflex replied:
yea. I'll fix any more if I see them.

momchil-flex commented on tidy3d/plugins/smatrix/smatrix.py:
```diff
@@ -92,32 +90,141 @@ class ComponentModeler(Tidy3dBaseModel):
         """Make sure simulation has no sources as they interfere with tool."""
         if len(val.sources) > 0:
-            raise SetupError("Simulation must not have sources.")
+            raise SetupError("Simulation must not have extraneous `sources`.")
```
'sources'... or just sources?

momchil-flex commented on tidy3d/plugins/smatrix/smatrix.py:
```diff
@@ -92,32 +90,141 @@ class ComponentModeler(Tidy3dBaseModel):
         None,
         title="Batch",
-        description="Batch of task used to compute S matrix. Set internally.",
+        description="Batch containing all of the simulations needed for the scattering matrix." "",
```
Shouldn't this say that it is set internally? I don't think there's a scenario where the user supplies it?

tylerflex replied:
I guess it's possible the user downloaded the batch separately from the component modeler and wants to supply it. But yea I'll mention to not use this unless you know what you're doing.

tylerflex commented:
Ok fixed all these comments. Thanks!

momchil-flex reviewed (approved):
Approved, just maybe rebase once again (no need to necessarily squash to single commit but more meaningful commit messages if not)
